src/hotspot/share/oops/instanceKlass.hpp

```diff
@@ -1080,6 +1080,12 @@ class InstanceKlass: public Klass {
   bool idnum_can_increment() const      { return has_been_redefined(); }
   inline jmethodID* methods_jmethod_ids_acquire() const;
   inline void release_set_methods_jmethod_ids(jmethodID* jmeths);
+  // Used to explicitly clear jmethodIDs for the contained methods
+  // This is required for JDK-8313816 but should not be used otherwise!
+  //   - We can not use the jmethodID cache associated with klass directly because the 'previous' versions
+  //     do not have the jmethodID cache filled in. Instead, we need to lookup jmethodID for each method and this
+  //     is expensive - O(n) for one jmethodID lookup. For all contained methods it is O(n^2).
```

Comment:
The comment is helpful, but I would move this part to the implementation. Here, what matters to know is that this function nulls out jmethodIDs for all methods in this IK.

Do I understand correctly that this comment tries to explain why we - instead of just iterating IK->_methods_jmethod_ids directly - we iterate all methods of this IK and then lookup the associated jmethodID in IK->_methods_jmethod_ids, right? I wondered about that but IIUC the pointers inside IK->_methods_jmethod_ids may refer to jmethodID slots that have been reused for different methods?

Reply:
>I wondered about that but IIUC the pointers inside IK->_methods_jmethod_ids may refer to jmethodID slots that have been reused for different methods?

Yes. The reason is that if a class has previous versions, these versions do not contain their own jmethodID cache but rather delegate to the 'main' version. So we have nothing to iterate over when a previous class version gets unloaded - also, to make things more interesting, the previous method versions are pointing to the main class version as their `method_holder`. Making it impossible to iterate over the full jmethodID cache in the main class version and nulling out a jmethodID only if the method it points to has `method_holder` pointing to the class being unloaded (it will always point to the main class version ...). 

If there is an easy way out of this without incurring O(n^2) complexity, please speak up. I was not able to figure out an alternative without eg. keeping an explicit method->jmethodID link but that would increase the Method instance footprint and it felt like a more invasive change for what I want to achieve.